README.md
**#Nescience Neural Network Classifier [WIP]**

There are two main versions of Nescience Neural Network Classifier:

+ NNNClassifier: Version using sequential (unoptimized) search. 

+ NNC-GE: Uses Grammatical Evolution (PonyGE2 algorithm) for developing the networks in a genetic-based way, focusing on minimizing the Nescience value on each generation.

**Folder structure:**

- NNC-GE/PonyGE2: GE Version of Nescience NNC. Relevant files:
    - `grammars/mlpGrammar.pybnf`: Contains model grammars in Backus-Naur Form
    - `parameters/mlpConfig.txt`: Config file setting Genetic Algorithm parameters (num generations, population size...)
    - `src/fitness/NescNNclasGE.py`: Contains main evaluation (fitness) and Nescience functions.
    - `src/ponyge.py`: Algorithm execution script.
    - `src/datasets.py`: Contains datasets and internal load functions
    - `src/usedata.py`: Dataset selection script. Creates usedata.txt file for internal use.
    - `src/nnncutils.py`: Contains Retrieval and Evaluation functions for post-run analysis

**Usage of NNNC-GE Version:**

- ***Inside e.g Google Colab notebook (Free GPU):*** **(all libraries included in Colab env)**
    - Look at template: https://colab.research.google.com/drive/1pJOq9CmIe7eCgdVlpB01-PFMBAzdry8F

- ***As a script:*** **(some libraries may be required)**
    
    (Inside PonyGE2/src folder) 
    
    - 1.[Selecting dataset]: `python usedata.py <dataset>`
    - 2.[Algorithm execution]: `python ponyge.py --verbose --parameters mlpConfig.txt --generations 10 --population_size 15`
    - 3.[Results]: `python nnncutils.py <dataset> <runtime>`


 >Where \<dataset\> is one of the following: [digits, fmnist, magic, pulsars, iris]
    and \<runtime\> is algorithm exec time (same format as in analysis/ folder) "hh:mm:ss"

**Results evaluation:**

The algorithm will generate and store every individual (network) in .h5 format inside `analysis/<runtime>/networks` folder, so that it can be later retrieved 
using keras `load_model()` function, obtaining the full structure with its weights resulted from training. The best 
network (achieving lowest Nescience) will be also stored and can be recovered for later use loading its .h5 file identified
with its individual number (shown in algorithm).

To ease a quick evaluation of the results, the `nnncutils.py` script retrieves the model and evaluates it on the test set.

> Note: All networks and parameter files resulting from the algorithm execution will be stored inside the `src/analysis/<runtime>/` folder.

`analysis/<runtime>/stats.csv` file contains a dataframe that stores each evaluated individual's Nescience parameters.

CHANGELOG
commit 6fe269fd9d4a0b4d461609cff208fcef7e88ee9d
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Fri Sep 13 09:47:35 2019 +0200

    Updated with new Nescience library

commit 1da135fece0a73127250c319b0967297642dc3f7
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Fri Sep 13 09:44:56 2019 +0200

    Updated NNC-GE algorithm to new Nescience library version

commit 65c9ad69052efab29690f8fcaec29be396ef81d3
Author: Andres Carrillo Lopez <acarrillolopez98@gmail.com>
Date:   Tue Sep 3 07:19:32 2019 +0000

    Updated README.md with easier to use Nescience NN AutoML Template

commit f43a57f5aee927b74b2ddfca2e0add36c8ec9241
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Mon Sep 2 10:48:33 2019 +0200

    Solved issue with load_weights after Keras 2.2.5 update

commit bd9d614e949635542a8c5768eae4a12b0d7f6db7
Author: Andres Carrillo Lopez <acarrillolopez98@gmail.com>
Date:   Mon Sep 2 07:15:01 2019 +0000

    Update README.md

commit b9b9c94e27c9f76ddd90d81a541703e77745878a
Author: Andres Carrillo Lopez <acarrillolopez98@gmail.com>
Date:   Wed Aug 7 07:10:07 2019 +0000

    Update README.md

commit 835a6b004c7e657992e1a090592716a3dee6c8e5
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Aug 6 15:30:20 2019 +0200

    Updated changelog

commit bb1848bbaeb8cd58f537e7f60c499a09dd4e7ed7
Author: Andres Carrillo Lopez <acarrillolopez98@gmail.com>
Date:   Tue Aug 6 13:25:41 2019 +0000

    Update README.md

commit 2d546694af4d6fa9d3c82627ddca9dd52ab2d6d9
Author: Andres Carrillo Lopez <acarrillolopez98@gmail.com>
Date:   Tue Aug 6 13:02:40 2019 +0000

    Update README.md

commit 0f2b4b4a0bc0a9d494908163ac485e481097b381
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Aug 6 14:27:13 2019 +0200

    Added script for evaluating algorithm result

commit f421259f99983cff68463d6551745e69630103f5
Merge: 95e5a08 8ad4a5d
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Aug 6 14:26:33 2019 +0200

    Merge branch 'master' of https://gitlab.com/andresC98/NescienceNNClassifier

commit 95e5a08923542122768880b2f42dbe061f8036f9
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Aug 6 13:23:09 2019 +0200

    Added NumVIUs metric to stats file for atomatic best model retrieval

commit 8ad4a5ddbdfa8f2255716aaed61f82d69a3a53d2
Author: Andres Carrillo Lopez <acarrillolopez98@gmail.com>
Date:   Tue Aug 6 10:34:36 2019 +0000

    Update README.md

commit 028499bb7c38afa8246baa27c9909edc0396a745
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Aug 6 11:39:00 2019 +0200

    Added new Dataset selection script. Working version

commit b6b8ea3e6110569fd1ac94f649bc687515b90ddc
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Mon Jul 29 08:35:28 2019 +0200

    Added random undersampling option. Added full model save and  ModelCheckpoints callbacks. Improved saved analysis files.

commit 16a746b7fd549b9ffa36f1c0578a390e667f9f15
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Thu Jul 18 11:17:08 2019 +0200

    Added variable-size output for grammar (compatible with any dataset). Added file-generation for post-run analysis

commit 8baa7c39bba03526b5c3ec6787b52e79423957f5
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Jul 16 14:39:33 2019 +0200

    Added inaccuracy temporal fix and miscoding to nsc stats

commit 7da362f32f49c9b7cc36351d10af387d18c9ff2c
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Jul 16 12:29:23 2019 +0200

    Added new VIU and EnhancedMiscoding into grammar

commit 17c826b2a18ce6545bff2162d1c8618b7ccc4da7
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Jul 16 09:58:59 2019 +0200

    Added .csv generation with Nescience stats for individuals

commit bd3d07654bb108070f2ef02c750728cd99126321
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Thu Jul 11 12:16:05 2019 +0200

    Removed debug prints from output

commit e2d11a65bd85c2b0f9ad38fdd8346b1a2b862c67
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Thu Jul 11 11:58:58 2019 +0200

    Solved critical misuse of Training/Test information for inaccuracy evaluation

commit 797e87b9040024aaa20771453b3c53af3fdd5897
Merge: 3beecd7 32c0050
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Wed Jul 10 15:10:58 2019 +0200

    Merge branch 'master' of https://gitlab.com/andresC98/NescienceNNClassifier

commit 3beecd7fcfa20df75c32d1e779c28acfda728a20
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Wed Jul 10 15:10:49 2019 +0200

    Simplified grammar choices for testing

commit 32c00508a8b37cfce14570a3da1ebe624123a761
Author: Andres Carrillo Lopez <acarrillolopez98@gmail.com>
Date:   Wed Jul 10 08:51:24 2019 +0000

    Added README.md

commit b07adb87307e575adffe7a32b994b222ebe3e953
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Wed Jul 10 10:44:09 2019 +0200

    Updated CHANGELOG

commit a99e26f44f5cc664a5143fd6c5fa86ac1f80616a
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Jul 9 16:34:04 2019 +0200

    Removed unused files

commit 89a469cd3acff0557a5f1c99a3649719b5383090
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Jul 9 16:10:15 2019 +0200

    Solved bottleneck in enhanced miscoding computation.

commit b81909b445700dde8c89e103dae5b45cab82e4d1
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Jul 9 12:54:52 2019 +0200

    Fixed issue. Working Enhanced Miscoding version.

commit e64c0884a4102d55c128ea265a4ea9b636f26354
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Tue Jul 9 12:29:16 2019 +0200

    added computation of optimal viu

commit bf72ee2798d86d56a4cde070a3f0b4d86878b6c3
Author: Andres Carrillo <acarrillolopez98@gmail.com>
Date:   Mon Jul 8 13:56:02 2019 +0200

    Alternative miscoding. Hid prints to make output cleaners.

commit 2adda4c98e902900e5875938247fea39b83833b0
Author: Andres Carrillo <acarrillolopez98@gmail.com>

AutoML.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Oct  9 14:07:07 2018

@author: rleiva
"""

# Generic
import random
import time

# Bokeh
from tornado        import gen
from bokeh.models   import ColumnDataSource, GraphRenderer, StaticLayoutProvider, Circle

from bokeh.plotting import curdoc, figure
from bokeh.layouts  import row, layout

# Treading
from functools      import partial
from threading      import Thread

# ML
import numpy as np
from sklearn.datasets import load_digits
from NNClassifier.NescienceNeuralNetworkClassifierV211 import *

count = 1

@gen.coroutine
def stream_update(x, vals):

    global f0
    
    nsc = vals["nescience"]
    mis = vals["miscoding"]
    sur = vals["surfeit"]
    ina = vals["inaccuracy"]
    scr = vals["score"]
    
    source_nsc.stream(dict(x=[x], y=[nsc]), rollover=100)
    source_mis.stream(dict(x=[x], y=[mis]), rollover=100)
    source_sur.stream(dict(x=[x], y=[sur]), rollover=100)
    source_ina.stream(dict(x=[x], y=[ina]), rollover=100)
    source_scr.stream(dict(x=[x], y=[scr]), rollover=100)

    # Display neural network
     
    canvas_size = 1
    top    = .9 * canvas_size
    bottom = .1 * canvas_size
    left   = .1 * canvas_size
    right  = .9 * canvas_size

    layer_sizes = vals["layer_sizes"]
    N = np.sum(layer_sizes)

    v_spacing = (top - bottom)/float(max(layer_sizes))
    h_spacing = (right - left)/float(len(layer_sizes) - 1)

    node_indices = list(range(N))

    graph = GraphRenderer()

    graph.node_renderer.data_source.add(node_indices, 'index')

    size = int(v_spacing * 300)
    graph.node_renderer.glyph = Circle(size=size, fill_color='white')

    # Plot edges
    
    start = list()
    end  = list()

    count = 0
    for n, (layer_size_a, layer_size_b) in enumerate(zip(layer_sizes[:-1], layer_sizes[1:])):
           
        for m in np.arange(layer_size_a):
            for o in np.arange(layer_size_b):
                start.append(count + m)
                end.append(count + layer_size_a + o)
    
        count = count + layer_size_a

    graph.edge_renderer.data_source.data = dict(start=start, end=end)
             
    # Nodes

    x = list()
    y = list()

    for n, layer_size in enumerate(layer_sizes):
    
        layer_top = v_spacing * (layer_size - 1)/2. + (top + bottom)/2.
        
        for m in np.arange(layer_size):

            x.append(n * h_spacing + left)
            y.append(layer_top - m * v_spacing)

    graph_layout = dict(zip(node_indices, zip(x, y)))
    graph.layout_provider = StaticLayoutProvider(graph_layout=graph_layout)

    if len(f0.renderers) != 0:
        f0.renderers.pop()
        
    f0.renderers.append(graph) 
     

def update_plot():
    
    global queue, count

    while True:
        
        vals = queue.get()
        queue.task_done()
        
        count = count + 1
                    
        doc.add_next_tick_callback(partial(stream_update, x=count, vals=vals))


# Data source for Bokeh
source_nsc = ColumnDataSource(data=dict(x=[0], y=[1]))
source_mis = ColumnDataSource(data=dict(x=[0], y=[1]))
source_sur = ColumnDataSource(data=dict(x=[0], y=[1]))
source_ina = ColumnDataSource(data=dict(x=[0], y=[1]))
source_scr = ColumnDataSource(data=dict(x=[0], y=[1]))

# Bokeh current session
doc = curdoc()
                
# Start web
f0 = figure(y_range=[0, 1], x_range=[0, 1], title="Neural Network", plot_width=300, plot_height=300, toolbar_location=None)

f1 = figure(y_range=[0,1], title="Nescience", plot_width=300, plot_height=300, toolbar_location=None)
f1.line(x='x', y='y', source=source_nsc)

f2 = figure(y_range=[0,1], title="Miscoding", plot_width=300, plot_height=300, toolbar_location=None)
f2.line(x='x', y='y', source=source_mis)

f3 = figure(y_range=[0,1], title="Inaccuracy", plot_width=300, plot_height=300, toolbar_location=None)
f3.line(x='x', y='y', source=source_ina)

f4 = figure(y_range=[0,1], title="Surfeit", plot_width=300, plot_height=300, toolbar_location=None)
f4.line(x='x', y='y', source=source_sur)

f5 = figure(y_range=[0,1], title="Validation Score", plot_width=300, plot_height=300, toolbar_location=None)
f5.line(x='x', y='y', source=source_scr)

l = layout([
    [f0],
    [f1, f2, f3, f4, f5]
    ], sizing_mode='stretch_both')

# doc.add_root(row(f1, f2, f3, f4, f5))
doc.add_root(l)

# Start web page thread
thread_consumer = Thread(target=update_plot)
thread_consumer.start()

# Prepare data and model

# scikit - Digits <----
# data = load_digits()
# X = data.data
# y = data.target
#alternative MNIST from keras:
# from keras.datasets import mnist
# (X_train, y_train), (X_test, y_test) = mnist.load_data()

#Pulsar Star Classification from Kaggle
from numpy import genfromtxt
data = genfromtxt('pulsar_stars.csv', delimiter=',')
data = np.delete(data, 0, 0)
X = data[:,:8]
y = data[:,8]

#Breast cancer
# from sklearn.datasets import load_breast_cancer
# data = load_breast_cancer()
# X = data.data
# y = data.target

# fashion_mnist = keras.datasets.fashion_mnist
# (train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()


# Numerai
# train = pd.read_csv('../data/numerai_training_data.csv', header=0)
# tournament = pd.read_csv('../data/numerai_tournament_data.csv', header=0)
# validation = tournament[tournament['data_type']=='validation']
# train_bernie = train.drop([
#     'id', 'era', 'data_type',
#     'target_charles', 'target_elizabeth',
#     'target_jordan', 'target_ken'], axis=1)
# features = [f for f in list(train_bernie) if "feature" in f]
# X = train_bernie[features]
# Y = train_bernie['target_bernie']
# X = np.array(X)
# y = np.array(Y)

# CIFAR-10
#import pickle
#
#with open("/data/CIFAR-10/cifar-10-batches-py/data_batch_1", 'rb') as fo:
#    data = pickle.load(fo, encoding='bytes')
#
#X = data[b"data"]
#y = data[b"labels"]

# Train the model

model = NescienceNeuralNetworkClassifier(backward=False, verbose=True)

thread_producer = Thread(target=model.fit, args=(X, y))
thread_producer.start()

